### PR TITLE
Serialization search order fix

### PR DESCRIPTION
**Pull request by asimarslan**
fixes https://github.com/hazelcast/hazelcast/issues/6455

### CODE REVIEW

**Comment by asimarslan** [comment]
verify

**Comment by sancar** [comment]
:+1:

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/SerializationServiceV1.java [hunk 1]
```diff
@@ -137,25 +137,24 @@ protected void registerConstantSerializers() {
...
+        //Java extensions: more serializers
+        registerConstant(Date.class, new DateSerializer());
+        registerConstant(BigInteger.class, new BigIntegerSerializer());
+        registerConstant(BigDecimal.class, new BigDecimalSerializer());
+        registerConstant(Class.class, new ClassSerializer());
+        registerConstant(Enum.class, new EnumSerializer());
+        registerConstant(ArrayList.class, new ArrayListStreamSerializer());
+        registerConstant(LinkedList.class, new LinkedListStreamSerializer());
+
+        safeRegister(Serializable.class, javaSerializerAdapter);
```
Comment (cangencer) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/SerializationServiceV1.java]: is this necessary to register Serializable? We already check if a class is Serializable in the lookupJavaSerializer, and means you can also skip removing the two interfaces in the lookupCustomSerializer

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/SerializationConstants.java [hunk 2]
```diff
@@ -63,29 +63,30 @@
...
 
-    public static final int JAVA_DEFAULT_TYPE_ARRAY_LIST = -107;
+    // NUMBER OF CONSTANT SERIALIZERS...
+    public static final int CONSTANT_SERIALIZERS_LENGTH = 28;
+
+    // ------------------------------------------------------------
+    // JAVA SERIALIZATION
 
-    public static final int JAVA_DEFAULT_TYPE_LINKED_LIST = -108;
+    public static final int JAVA_DEFAULT_TYPE_SERIALIZABLE = -100;
```
Comment (cangencer) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/SerializationConstants.java]: could we add default type of other languages here ? C# will be -101.

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/SerializationServiceV1.java (hunk 1 above)
Comment (asimarslan) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/SerializationServiceV1.java]: It is required for first time deserialization

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/SerializationConstants.java (hunk 2 above)
Comment (asimarslan) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/SerializationConstants.java]: We can add it as a comment as we don't have a serializer for it

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/SerializationServiceV1.java [hunk 3]
```diff
@@ -137,25 +137,24 @@ protected void registerConstantSerializers() {
...
-        safeRegister(Externalizable.class, new ExternalizableSerializer(enableCompression));
-        safeRegister(Serializable.class, new JavaSerializer(enableSharedObject, enableCompression));
-        safeRegister(Class.class, new ClassSerializer());
-        safeRegister(Enum.class, new EnumSerializer());
-        safeRegister(ArrayList.class, new ArrayListStreamSerializer());
-        safeRegister(LinkedList.class, new LinkedListStreamSerializer());
+    public void registerJavaTypeSerializers() {
+        //Java extensions: more serializers
+        registerConstant(Date.class, new DateSerializer());
+        registerConstant(BigInteger.class, new BigIntegerSerializer());
```
Comment (asimarslan) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/SerializationServiceV1.java]: @pveentjer backport point https://github.com/hazelcast/hazelcast/issues/6759#issuecomment-158191317

**Pull request merged**